test/checkpoint/checkpointer.c

```diff
@@ -373,6 +373,122 @@ real_checkpointer(THREAD_DATA *td)
+ *     Claim pending prepared transactions, do a checkpoint then verify the consistency of the data
+ *     post claiming.
+ */
+void
+prepare_discover(WT_CONNECTION *conn, THREAD_DATA *td)
+{
+    WT_UNUSED(td);
+    /*
+     * Since RTS is not ran with precise checkpoint, we need to use prepare discover cursor to claim
+     * all pending prepared transactions. When precise checkpoint is not configure, there's no need
+     * to run prepare discover.
+     */
+    if (!g.precise_checkpoint)
+        return;
+
+    WT_CURSOR *cursor;
+    WT_DECL_RET;
+    WT_SESSION *session;
+    uint64_t prepared_id;
+    uint32_t discover_count;
+    char buf[128];
+    char timestamp_buf[64];
+    bool should_commit;
+
+    testutil_check(conn->open_session(conn, NULL, NULL, &session));
+    /* Open the prepare discover cursor */
+    ret = session->open_cursor(session, "prepared_discover:", NULL, NULL, &cursor);
+    if (ret == WT_NOTFOUND) {
+        /* No prepared transactions found - this is normal */
+        testutil_check(session->close(session, NULL));
+        return;
+    }
+    testutil_check(ret);
+    /* Iterate through all prepared transactions and claim pending prepared transactions. */
+    discover_count = 0;
+    testutil_check(g.conn->query_timestamp(g.conn, timestamp_buf, "get=stable_timestamp"));
+    uint64_t stable_ts = testutil_timestamp_parse(timestamp_buf);
+    printf(
+      "stable_ts=%" PRIu64 ", latest_prepared_ts=%" PRIu64 "\n", stable_ts, g.latest_prepared_ts);
+    if (g.latest_prepared_ts >= stable_ts) {
+        stable_ts = g.latest_prepared_ts + 1;
+        g.ts_stable = stable_ts;
+        set_stable(stable_ts);
+    }
+    while ((ret = cursor->next(cursor)) == 0) {
+        uint64_t commit_ts, durable_ts, rollback_ts;
+
+        discover_count++;
+        testutil_check(cursor->get_key(cursor, &prepared_id));
+
+        /* Claim the prepared transaction */
+        testutil_snprintf(buf, sizeof(buf), "claim_prepared_id=%" PRIx64, prepared_id);
+        testutil_check(session->begin_transaction(session, buf));
+
+        /* Randomly decide whether to commit or roll back */
+        u_int rnd = __wt_random(&td->data_rnd);
+        should_commit = rnd % 2 == 0;
+
+        if (should_commit) {
+            printf("committing txn with prepared_id=%" PRIu64, prepared_id);
+            /* Get current stable timestamp and use it for new timestamps */
+            testutil_check(g.conn->query_timestamp(g.conn, timestamp_buf, "get=stable_timestamp"));
+            uint64_t current_stable = testutil_timestamp_parse(timestamp_buf);
+
+            /* Use timestamps greater than current stable */
+            commit_ts = current_stable + 1;
+            durable_ts = commit_ts + 2;
+
+            testutil_snprintf(buf, sizeof(buf),
+              "durable_timestamp=%" PRIx64 ",commit_timestamp=%" PRIx64, durable_ts, commit_ts);
+            testutil_check(session->commit_transaction(session, buf));
+
+            printf(
```

Comment:
Do we have printing support in test/checkpoint? We shouldn't print this every time I think.

Reply:
removed